Deadletter documentation added (#281)
AxonIQ/reference-guide: axon-framework/events/event-processors/README.md

> processing speeds in general.
>
> In all, you are free to use an ordering, but we recommend using it sparingly.

## Error Handling

Errors are inevitable in any application.
Depending on where they happen, you may want to respond differently.

By default, exceptions raised by event handlers are caught in
the [Processing Group layer](#processing-group-listener-invocation-error-handler), logged, and processing continues with
the following events.
When an exception is thrown when a processor is trying to commit a transaction, update
a [token](streaming.md#token-store), or in any other part of the process, the exception will be propagated.

In the case of a [Streaming Event Processor](streaming.md#error-mode), this means the processor will go into error mode,
releasing any tokens and retrying at an incremental interval \(starting at 1 second, up to max 60 seconds\).
A [Subscribing Event Processor](subscribing.md#error-mode) will report a publication error to the component that
provided the event.

To change this behavior, both the Processing Group and Event Processor level allow customization on how to deal with

> exceptions:

Author: I added the sentence `Based on the provided `ErrorContext` object, you can decide to ignore the error, schedule retries, perform
dead-letter-queue delivery, or rethrow the exception.`

Based on the provided `ErrorContext` object, you can decide to ignore the error, schedule retries, perform dead-letter-queue delivery, or rethrow the exception.

### Dead-Letter Queue

Although configuring a [Listener Invocation Error Handler](#processing-group---listener-invocation-error-handler)
and [Error Handler](#event-processor---error-handler)

> helps with dealing with event handling exceptions, you still end up in an event handling stop.

Reviewer: ```suggestion
helps you to deal with exceptions when processing events, you still end up in an event handling stop.
```

A perfect example when this is a requirement is the need to handle aggregate events in publishing order.
Simply dead lettering one failed event would cause subsequent events in the sequence to react to stale state.

It is thus of utmost importance that a dead-letter queue for events enqueues an event and any following events in the sequence.
To that end, the supported dead-letter queue is a so-called `SequencedDeadLetterQueue`.

Integral to its design is to allow for queueing failed events and events that belong to a faulty sequence.
It does so by maintaining a sequence identifier for each event, determined by the [sequencing policy](/axon-framework/events/event-processors/streaming.md#sequential-processing).

Note that you *cannot* share a dead-letter queue between different processing groups.
Hence, each processing group you want to enable this behavior for should receive a unique dead-letter queue instance.

We currently provide the following dead-letter queue implementations:
* `InMemorySequencedDeadLetterQueue` - In-memory variant of the dead-letter queue. 

>   Useful for testing purposes, but as it does not persist dead letters, it is unsuited for most production environments.

Author: most or any?

Reviewer 2: You are right. I've removed `most` entirely from the sentence.

}
```

Based on the provided `ErrorContext` object, you can decide to ignore the error, schedule retries, perform
dead-letter-queue delivery, or rethrow the exception.

### Dead-letter Queue

When event processor transactions end up in an exception, following events are not handled by that event processor even
though they could be successfully handled. The event processor is stuck until the issue is fixed.
To skip and save the events that are failing, you can configure a dead-letter queue for an event processor. A
dead-letter queue can not be shared between event processors so there should one queue per processor.
The `InMemorySequencedDeadLetterQueue` can be used for testing purposes but the dead-letters are gone after a restart.
To persist dead-letters the `JpaSequencedDeadLetterQueue` should be used.
When using the `JpaSequencedDeadLetterQueue` the dead-lettered events are stored in the `dead_letter_entry` database
table.

> 

Reviewer 2: Enter section:
```suggestion
#### Configuring a sequenced Dead-Letter Queue
```

                                                                    .entityManagerProvider(
                                                                            configuration.getComponent(
                                                                                    EntityManagerProvider.class))
                                                                    .serializer(
                                                                            configuration.serializer())
                                                                    .maxSequences(256)
                                                                    .maxSequenceSize(256)
                                                                    .build()));
    }
}
```
{% endtab %}

{% endtabs %}
You can set the maximum amount of sequences that are saved (defaults to 128) and the maximum amount of dead-letters in a
sequence (also defaults to 128). If these thresholds are exceeded an exception will be thrown and the event processor
will stop processing.
The sequence id is determined by the sequencing policy. By default, the sequence identifier is the aggregate id.

> 

Reviewer 2: Enter section:
```suggestion
#### Processing Dead-Letter Sequences
```

After fixing the issue the events can be handled again by using the `process` function, in this case it will process the
first matching event of type ErrorEvent:

{% tabs %}
{% tab title="Process first dead-letter in the queue of type ErrorEvent " %}

```java
public class DeadletterProcessor {

    private final EventProcessingConfiguration eventProcessingConfiguration;

    public DeadletterProcessor(EventProcessingConfiguration eventProcessingConfiguration) {
        this.eventProcessingConfiguration = eventProcessingConfiguration;
    }

>     public void repairErrorEvent() {

Reviewer 2: Technically speaking, there's no guarantee it'll be repaired. That depends on whether the handler process really is fixed.
Hence I think it's better to call this method `retryErrorEventSequence`.

        eventProcessingConfiguration.sequencedDeadLetterProcessor(PROCESSING_GROUP)
                                    .ifPresent(p -> p.process(deadLetter -> deadLetter.message()
                                                                                      .getPayload() instanceof ErrorEvent));
    }
}
```

{% endtab %}

> {% tab title="Process any dead-letter in the queue" %}

Reviewer 2: Now that I've typed this, perhaps we should call the method `processOldest` instead of `processAny`... What do you think?

        Optional<Iterable<Iterable<DeadLetter<? extends EventMessage<?>>>>> deadLetters = eventProcessingConfiguration.deadLetterQueue(
                PROCESSING_GROUP).map(SequencedDeadLetterQueue::deadLetters);

        deadLetters.ifPresent(sequences -> {
            // Iterate over all sequences
            for (Iterable<DeadLetter<? extends EventMessage<?>>> sequence : sequences) {
                // Iterate over all dead-letters belonging to the same sequence
                for (DeadLetter<? extends EventMessage<?>> deadLetterInSequence : sequence) {
                    eventProcessingConfiguration.sequencedDeadLetterProcessor(PROCESSING_GROUP).ifPresent(
                            eventMessageSequencedDeadLetterProcessor -> {
                                eventMessageSequencedDeadLetterProcessor.process(
                                        deadLetter -> deadLetter.message().getIdentifier()
                                                                .equals(deadLetterInSequence.message()
                                                                                            .getIdentifier()));
                            });

>                 }

Reviewer 2: I think this second for-loop isn't necessary. 
I intended the name of the `SequencedDeadLetterProcessor` to clarify  it processes dead-letter sequences.
Hence, every `process(Predicate)` invocation will already process the sequence.

As such, only the first entry of every sequence should be used to have all dead-letter sequences be processed.

            }
        });
    }
}
```

{% endtab %}
{% endtabs %}

> 

Reviewer 2: Enter section:
```suggestion
#### Dead-Letter Enqueue Policy
```
I am least certain about this section header.
Perhaps "Deciding when to Enqueue" is clearer to the reader.

> You can implement a custom dead-letter policy to exclude some events from the dead-letter queue, these events will be

Reviewer 2: I would expect a form of introduction sentence, stating something like "by default all event handling failures result in enqueueing the event as a dead-letter".
Followed up by something like "but this might not be desirable for every failure".
Then, the intro section could end with "you can impact this decision process by configuring your own `EnqueuePolicy`.

> skipped. This policy is not only called for initial failures but also when dead-lettered events are processed

Reviewer 2: A second paragraph would indeed emphasize that the policy also dictates what to do when retrying fails.
Thus containing something like "this allows you to implement, for example, a maximum amount of retries, by adding diagnostic information like the number of retries to the dead-letter".

unsuccessfully again.

```java
@Configuration
public class CustomDeadLetterPolicy {

    @Autowired
    public void configure(EventProcessingConfigurer configurer) {
        configurer.registerDeadLetterPolicy(PROCESSING_GROUP, configuration ->
                (letter, cause) -> {
                    if (cause instanceof NullPointerException) {
                        // It's pointless..
                        return Decisions.doNotEnqueue();
                    }

>                     final int retries = (int) letter.diagnostics().getOrDefault("retries", -1);

Reviewer 2: The `diagnostics` bit makes me wonder whether we should have another subsection about what the `DeadLetter` stores.
The section would be short and simple I think, essentially a version of the API docs of the `DeadLetter`.
Thus, a bullet-list of the fields and what there intent is. 

That way, we do not have to explain diagnostics in this section, as this section should be focused around how to make that decision.

                    if (letter.message().getPayload() instanceof ErrorEvent) {
                        // Important, always retry
                        return Decisions.enqueue(cause);
                    }
                    if (retries < 10) {
                        // Let's continue and increase retries!
                        return Decisions.enqueue(cause,
                                                 deadLetter -> deadLetter.diagnostics().and("retries", retries + 1));
                    }
                    // Exhausted retries
                    return Decisions.doNotEnqueue();
                });
    }
}

```

> One important note, when implementing event handlers, make them idempotent. With the dead-letter queue this is a hard requirement.

Reviewer 2: Awesome, we definitely need this paragraph indeed.
I think it would be good if we had this as a "notice section".

The notion for that is like how you'd do references.
The `> **Ordering Considerations*` bit on line 155 is a sample of this.

Furthermore, I'd like this notice to come earlier within the DLQ section.
Perhaps even directly after the introduction of this section.
That way, readers know there handlers should be idempotent "before" they've already read everything else.
